frontends/concrete-python/concrete/fhe/extensions/synthesis/verilog_to_luts.py

```diff
@@ -0,0 +1,346 @@
+"""
+Rewrite yosys json output as a simple Tlu Dag.
+"""
+import json
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+
+VERBOSE = False
+
+
+def log(*args):
+    """Log function statically activated."""
+    if VERBOSE:
+        print(*args)
+
+
+## Assuming same arity everywhere, i.e. dag with uniform precision TLUs
+LUT_COSTS = {
+    1: 29,
+    2: 33,
+    3: 45,
+    4: 74,
+    5: 101,
+    6: 231,
+    7: 535,
+    8: 1721,
+}
+
+
+def luts_spec_abc():
+    """Generate the costs table (area, delay) for `abc`."""
+    return "\n".join(
+        # arity area delay
+        f"{arity}\t{cost}\t{cost}"
+        for arity, cost in LUT_COSTS.items()
+    )
+
+
+YOSIS_EXE_NAME = "yowasp-yosys"
+_yosis_exe = None  # pylint: disable=invalid-name
+
+
+def detect_yosis_exe():
+    """Detect yosis executable."""
+    global _yosis_exe  # noqa: PLW0603 pylint: disable=global-statement
+    if _yosis_exe:
+        return _yosis_exe
+    result = (
+        shutil.which(YOSIS_EXE_NAME)
+        or shutil.which(YOSIS_EXE_NAME, path=os.path.dirname(sys.executable))
+        or shutil.which(YOSIS_EXE_NAME, path=os.path.dirname(shutil.which("python3") or ""))
+    )
+    if result is None:
+        msg = f"{YOSIS_EXE_NAME} cannot be found."
+        raise RuntimeError(msg)
+    _yosis_exe = result
+    return result
+
+
+def yosys_script(abc_path, verilog_path, json_path, dot_file, no_clean_up=False):
+    """Generate `yosys` scripts."""
+    no_cleanup = "-nocleanup -showtmp" if no_clean_up else ""
+    return f"""
+echo on
+read -sv {verilog_path};
+prep
+techmap
+log Synthesis with ABC: {abc_path}
+abc {no_cleanup} -script {abc_path}
+write_json {json_path}
+""" + (
+        "" if not dot_file else "show -stretch"
+    )
+
+
+def abc_script(lut_cost_path):
+    """Generate `abc` scripts."""
+    return f"""
+# & avoid a bug when cleaning tmp
+read_lut {lut_cost_path}
+print_lut
+strash
+&get -n
+&fraig -x
+&put
+scorr
+dc2
+dretime
+strash
+dch -f
+if
+mfs2
+#lutpack
+"""
```

Comment:
Some documentation would be nice :sweat_smile:

Reply:
I have noclear  idea what theses lines do, it's the default abc script (Ill check), I just made it explicit to be able to modify it.